### PR TITLE
feat(speculative): add DSpark draft for MiniMax M3 VL (text + multimodal)

### PR DESCRIPTION
**Pull request by khazic**
## Summary

Adds a DSpark speculative-decoding draft for the MiniMax M3 VL target, mirroring the pattern already used for Qwen3 / Gemma4 / DeepSeek V4 (#2866). MiniMax M3 VL's base model is already merged (#2538 + follow-ups), so this PR is purely the DSpark draft side.

- `draft_minimax_m3.py`: a small dense GQA draft (separate q/k/v/o, per-head Gemma RMSNorm, SwiGLU-OAI, partial RoPE via the gpt_oss rotary utilities matching the target's own convention). Ignores M3's 128-expert MoE and block-sparse indexer attention entirely, same "draft is always dense" convention as every other DSpark target. Pinned to uniform fp32 storage at the end of `__init__` for FSDP2's uniform-dtype requirement (same defensive guard the DeepSeek V4 draft uses).
- `build_minimax_m3_draft_config` in `dspark/config.py`, following the same structure as `build_gemma4_draft_config`/`build_deepseek_v4_draft_config`.
- `train_dspark.py`: loads the ~400B-parameter MoE VLM target frozen through the same expert-parallel/FSDP distributed path the VLM finetune recipe uses (`create_distributed_setup_from_config` + `NeMoAutoModelForImageTextToText.from_pretrained`). `pp_size` must stay 1 since DSpark's forward-hook hidden-state capture needs one non-pipelined call.
- Opt-in multimodal (image+text) training support (`recipe_args.multimodal: true`): a dedicated collate function (`dspark_vlm_collate_fn`) reuses the VLM pipeline's tokenization/label-marking building blocks but stops before their shift-and-truncate step, since DSpark's own anchor sampling expects unshifted, equal-length `input_ids`/`loss_mask` (unlike `default_collate_fn`'s standard next-token-prediction shift). Text-only examples in a multimodal batch get the same fake-placeholder-image injection `default_collate_fn` already uses, so mixed text/image corpora stay FSDP-safe (M3's vision_tower is its own sharded unit; skipping it on some ranks but not others would desync the all-gather and hang training).
- `HFDSparkTargetModel.generate_batch()` now forwards any multimodal kwargs present in the batch (filtered against the target's own forward signature via `filter_forward_kwargs`), a no-op extension for every existing text-only caller (Qwen3, Gemma4).

## What is and is not verified

CPU unit tests pass (draft forward shapes, RoPE device-pinning, config builder, the multimodal collate function's unshifted-length contract and fake-image injection, `generate_batch`'s kwarg forwarding). I re-checked this implementation against the already-merged DeepSeek V4 DSpark draft (#2866) for structural consistency and found no repeated mistakes.

I do not have a MiniMax M3 checkpoint downloaded yet (854 GB, still in progress), so the full distributed target load, real EP/FSDP sharding, and actual multimodal data through the target have **not** been run on real hardware. @HuiyingLi, could you help validate on GPUs once you have a chance? Specifically:

1. `NeMoAutoModelForImageTextToText.from_pretrained(..., distributed_setup=..., backend=BackendConfig(attn="sdpa", linear="torch", experts="gmm", dispatcher="hybridep", ...), text_config={"num_mtp_modules": 0})` actually loads and shards the real checkpoint correctly.
2. `pp_size=1` with a larger `ep_size` (128 to match the validated VLM SFT yaml's per-rank expert memory, or less since DSpark's target is forward-only with no optimizer state) fits without OOM.
3. The captured hidden states / target forward are finite with a real multimodal batch (`pixel_values`/`image_grid_thw` actually reaching the target correctly).
4. The fake-image injection path for text-only samples in a mixed corpus does not hang FSDP2 training (this is the main new risk this PR introduces).

**1. Unit tests (CPU, ~15s):**

```
python -m pytest tests/unit_tests/speculative/test_dspark_minimax_m3.py tests/unit_tests/datasets/vlm/test_dspark_collate.py tests/unit_tests/speculative/test_dspark_target.py -q
```

**2. Text-only training:**

```
torchrun --standalone --nproc_per_node=8 -m nemo_automodel.recipes.llm.train_dspark -c examples/speculative/dspark/minimax_m3_dspark.yaml
```

**3. Multimodal training:**

```
torchrun --standalone --nproc_per_node=8 -m nemo_automodel.recipes.llm.train_dspark -c examples/speculative/dspark/minimax_m3_dspark_multimodal.yaml
```

Both yamls need `target_model_name_or_path` pointed at a local checkpoint and `train_data_path`/`dataset:` pointed at real (target-regenerated) data. Thanks in advance for testing.

## Test plan

- [x] `ruff format` / `ruff check` clean
- [x] `lint-imports` (components-must-not-import-each-other contract) clean
- [x] CPU unit tests added/passing (draft forward shapes, config builder, RoPE device pin, multimodal collate contract, fake-image injection, `generate_batch` kwarg forwarding)
- [ ] Real-GPU validation (target load, EP/FSDP sharding, multimodal data through the target) -- pending @HuiyingLi

### CODE REVIEW

**Comment by HuiyingLi** [comment]
@khazic I ran the first real-checkpoint text-only smoke on 8 GPUs at PR head `74111d0a`.

### Setup

- 1 node, 8x H100 80GB
- NVIDIA PyTorch 26.04 container, PyTorch `2.12.0a0`, CUDA 13.2
- Local full BF16 `MiniMaxAI/MiniMax-M3` checkpoint
- FSDP2: `tp=1, cp=1, pp=1, ep=8`
- Target backends: `attn=sdpa`, `dispatcher=hybridep`, `experts=torch_mm`
- Diagnostic target depth: first 4/60 text layers
- Captured target layers: `[1, 2, 3]`
- Draft depth: 2 layers
- `seq_length=512`, micro-batch size 1, 8 text samples, 1 optimizer step

Layer 3 is the first MiniMax M3 MoE + sparse-attention layer, so this reduced run still exercises real checkpoint loading/sharding, HybridEP dispatch, `torch_mm` experts, sparse target attention, hidden-state capture, draft forward/backward, and an optimizer update.

For this one-node diagnostic I locally made only the following `train_dspark.py` setup changes:

1. Applied the existing `recipe_args.target_num_hidden_layers` diagnostic override to `target_config.text_config.num_hidden_layers` before MiniMax target construction.
2. Made the MiniMax target expert backend configurable and selected `target_experts: torch_mm`.

Launch:

```bash
torchrun --standalone --nproc_per_node=8 \
  -m nemo_automodel.recipes.llm.train_dspark \
  -c minimax_m3_smoke_4layers.yaml
```

### Result

The run exited successfully after one optimizer step with no OOM or NCCL hang:

```text
step 1 | epoch 0 | loss 3.4655 | ce 13.6188 | tv 1.3961 |
conf 0.8472 | lr 6.00e-04 | mem 22.46 GiB
Finished epoch 1/1 completed_steps=1
```

The reported CUDA peak allocation was 22.46 GiB on rank 0. The training step took 25.9 seconds after setup/loading.

This is a successful reduced-target plumbing smoke only. It does **not** validate the complete 60-layer target, the intended capture points `[10, 22, 34, 46, 58]`, full-training memory requirements, or the multimodal data path.

**Comment by HuiyingLi** [comment]
@khazic I pushed commit `cc42df7` directly to this PR with the full-target experiment setup and the fixes found while moving from the 8-GPU diagnostic to real 16-node training.

## Committed fixes

### 1. MiniMax target setup controls

- Apply the existing diagnostic `recipe_args.target_num_hidden_layers` control to MiniMax's nested `text_config.num_hidden_layers` before model construction. This is explicitly diagnostic/CI-only; the full experiment leaves it unset and loads all 60 layers.
- Make the MiniMax target expert backend configurable with `recipe_args.target_experts`. The validated run uses `torch_mm`; target dispatch remains `hybridep`.

### 2. Optimizer construction and observability

- Preserve a dotted optimizer `_target_` string when the configuration is a real `ConfigNode`. `to_dict()` resolves it to a Python class, but the optimizer builder requires a registry/import-path string.
- Log the instantiated optimizer and master/moment settings at startup. The validated setup confirms TE `FusedAdam`, constant LR `1e-5`, FP32 master weights, parameter remainders, and FP32 moments.

### 3. Compile MiniMax draft FlexAttention

- The MiniMax draft called `flex_attention` eagerly. At `seq_length=4096`, PyTorch warned that this materializes the full score matrix and then OOMed in `math_attention`.
- Add a lazy CUDA `torch.compile(flex_attention, mode="max-autotune-no-cudagraphs", dynamic=True)` path, while retaining eager behavior for CPU/unsupported test shapes.

### 4. Bound full-vocabulary TV/confidence-loss memory

- The old loss path independently materialized full FP32 draft/target softmax tensors for both TV loss and confidence targets.
- MiniMax's shape is `512 anchors * 7 positions * 200064 vocab`; one FP32 probability tensor is about 2.67 GiB per rank. The second materialization OOMed even with EP128.
- Compute the same FP32 probability L1 distance exactly in 128-token chunks, activation-checkpoint each chunk for backward, and reuse the result for the confidence target. This changes memory behavior, not the objective or gradient.

### 5. Validated experiment config

Added `examples/speculative/dspark/minimax_m3_dspark_perfectblend_exp.yaml`:

- Direct Open-PerfectBlend input; no response regeneration.
- Full 60-layer BF16 MiniMax-M3 target, capture layers `[10, 22, 34, 46, 58]`.
- 16 nodes / 128 H100 80GB, FSDP2, `tp=1`, `cp=1`, `pp=1`, `ep=128`.
- HybridEP target dispatch, `torch_mm` experts, sequence length 4096, 512 anchors, five draft layers.
- Microbatch 1, accumulation 4, effective global batch 512.
- TE FusedAdam with constant LR `1e-5` and FP32 master weights.
- Checkpoint every 25 steps and W&B enabled.

The launch also gives each torchrun rank a private node-local Triton cache. This is a cluster-launch mitigation, not a model-code change; sharing `/root/.triton/cache` across nodes produced stale handles and partially-written CUBIN files.

## Validation progression

1. **8-GPU reduced smoke:** first 4/60 target layers, including the first MoE layer; one finite optimizer step completed. This was plumbing validation only.
2. **Job `13302722`, 8 nodes / EP64:** reached the first target forward, then ranks collided in the shared Triton cache (`Stale file handle` / missing `_indices_to_multihot_kernel.cubin`).
3. **Job `13302790`, 8 nodes / EP64, private caches:** passed target loading but eager FlexAttention OOMed while allocating the dense score/softmax matrix.
4. **Job `13303003`, 16 nodes / EP128, compiled FlexAttention:** passed target and draft attention, then the duplicated full-vocabulary FP32 softmax OOMed on a 2.67 GiB allocation.
5. **Job `13303127`, 16 nodes / EP128, all fixes:** currently training normally. Snapshot at step 77; checkpoints at steps 25, 50, and 75.

Current run loss samples:

```text
step 1  | loss 3.7423 | ce 13.1223 | tv 1.9985 | conf 0.6513 | mem 55.83 GiB
step 10 | loss 2.7988 | ce  8.5713 | tv 1.9120 | conf 0.2107 | mem 55.91 GiB
step 25 | loss 2.6754 | ce  7.6716 | tv 1.9332 | conf 0.1590 | mem 55.91 GiB
step 50 | loss 2.6692 | ce  7.5163 | tv 1.8837 | conf 0.2132 | mem 55.91 GiB
step 75 | loss 2.6044 | ce  6.8118 | tv 1.8400 | conf 0.2504 | mem 55.91 GiB
step 77 | loss 2.5690 | ce  6.5488 | tv 1.8500 | conf 0.2307 | mem 55.91 GiB
```

After first-step JIT, steady-state throughput is approximately 15 seconds/optimizer step. This validates full-target execution and checkpointing; it is not a convergence claim.

## W&B

- Current fixed run: https://wandb.ai/Nemo-automodel/minimax-m3-dspark/runs/5t181cj2
- EP128 pre-loss-fix run: https://wandb.ai/Nemo-automodel/minimax-m3-dspark/runs/2o104lah
- EP64 eager-FlexAttention run: https://wandb.ai/Nemo-automodel/minimax-m3-dspark/runs/t5dlm01s
- Initial shared-cache run: https://wandb.ai/Nemo-automodel/minimax-m3-dspark/runs/em4b7fql

## Existing raw logs

Shared cluster paths:

```text
/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_nemofw/users/huiyingl/sd/test-runs/pr2877-minimax-m3-smoke-4layers/smoke-rerun.log

/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_nemofw/users/huiyingl/sd/test-runs/pr2877-minimax-m3-perfectblend-full/logs/m3-dspark-pb-full-13302722.{out,err}
/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_nemofw/users/huiyingl/sd/test-runs/pr2877-minimax-m3-perfectblend-full/logs/m3-dspark-pb-full-13302790.{out,err}
/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_nemofw/users/huiyingl/sd/test-runs/pr2877-minimax-m3-perfectblend-full/logs/m3-dspark-pb-full-13303003.{out,err}
/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_nemofw/users/huiyingl/sd/test-runs/pr2877-minimax-m3-perfectblend-full/logs/m3-dspark-pb-full-13303127.{out,err}
```

## Checks

- `48 passed` across the focused DSpark methodology, MiniMax draft, and recipe-helper suites.
- Ruff passed.
- `git diff --check` passed.
- The new experiment YAML was parsed and checked for EP128, PP1, full target depth, `torch_mm`, accumulation 4, and TE FP32-master settings.

I did not add the state-dict-adapter bookkeeping workaround. The existing `missing=114, unexpected=0` grouped-expert warning remains warning-only; checkpoint loading and full training proceed correctly.

**Review by HuiyingLi — DISMISSED** [review]

**Review by HuiyingLi — APPROVED** [review]
